haskell/test/SMT/Common/Run.hs

```diff
@@ -0,0 +1,263 @@
+-- SPDX-FileCopyrightText: 2021 TQ Tezos
+-- SPDX-License-Identifier: LicenseRef-MIT-TQ
+
+module SMT.Common.Run
+  ( runBaseDaoSMT
+  ) where
+
+import Universum hiding (drop, swap)
+
+import qualified Data.Map as Map
+import Data.Sequence (Seq(..))
+import Fmt (pretty)
+import Hedgehog
+
+import Lorentz hiding (now, (>>))
+import qualified Lorentz.Contracts.Spec.FA2Interface as FA2
+import Lorentz.Test.Consumer (contractConsumer)
+import Michelson.Typed.Convert (convertContract, untypeValue)
+import Michelson.Untyped.Entrypoints (unsafeBuildEpName)
+import Morley.Nettest
+import qualified Morley.Micheline as MM
+import Morley.Nettest.Pure
+
+import Ligo.BaseDAO.Contract
+import Ligo.BaseDAO.Types
+import SMT.Model.BaseDAO.Contract
+import SMT.Model.BaseDAO.Types
+import SMT.Common.Gen
+import SMT.Common.Types
+import Test.Ligo.BaseDAO.Common.StorageHelper as StorageHelper
+import Test.Ligo.BaseDAO.Common
+
+-- | The functions run the generator to get a list of entrypoints
+-- and do setups before calling entrypoints againts ligo and haskell model
+-- For Ligo:
+--   - Setup initial level (generated by the generator)
+--   - Originate auxiliary contracts
+--   - Originate basedao contract
+-- For Haskell:
+--   - Setup `ModelState`
+runBaseDaoSMT :: SmtOption -> PropertyT IO ()
+runBaseDaoSMT option@SmtOption{..} = do
+
+  -- Run the generator to get a function that will generate a list of entrypoint calls.
+  mkModelInput <- forAll (runGeneratorT (genMkModelInput option) $ initGeneratorState soMkPropose)
+
+  nettestTestProp $
+    (uncapsNettestEmulated $ do
+        -- Originate auxiliary contracts
+        guardianContract <- originateSimple "guardian" () dummyGuardianContract
+        tokenContract <- originateSimple "TokenContract" [] dummyFA2Contract
+        registryDaoConsumer <- originateSimple "registryDaoConsumer" []
+          (contractConsumer @(MText, (Maybe MText))) -- Used in registry dao.
+
+        -- Generate a list of entrypoint calls
+        let ModelInput (contractCalls, ms) = mkModelInput $ ModelInputArg
+              { miaGuardianAddr = unTAddress guardianContract
+              , miaGovAddr = unTAddress tokenContract
+              , miaViewContractAddr = registryDaoConsumer
+              }
+
+        -- Set initial level for the Nettest
+        advanceLevel (ms & msLevel)
+
+        -- Modify `FullStorage` from the generator with registry/treasury configuration.
+        let newMs = ms { msFullStorage = soModifyFs (ms & msFullStorage) }
+
+        -- Originate Dao for Nettest
+        let
+          originateData = UntypedOriginateData
+            { uodName = "BaseDAO"
+            , uodBalance = toMutez 0
+            , uodStorage = untypeValue $ toVal $ (newMs & msFullStorage)
+            , uodContract = convertContract baseDAOContractLigo
+            }
+
+        daoUntyped <- originateUntyped originateData
```

Comment:
Can we not use `originateTypedSimple` here?

Reply:
Updated.

Reply:
My broader point was that I don't think we need `daoUntyped` at all. We should be able to pass the typed address `dao` around.

Reply:
Yeah, but we're interested in both `TAddress` and `Address`. Example: `msSelfAddress` is an `Address` because it uses as a part of `FA2.TransferItem` `tiFrom`. So even if we use `TAddress` as the type of `msSelfAddress`, we still have to convert it to `Address` at the end.

Reply:
We can use `toAddress` at the point where we need an actual `Address` (like the example you mentioned, when creating a `FA2.TransferItem`). But otherwise, I think we should keep the type safety on (e.g. when we pass it to `sendXtzWithAmount` and to `TransferData`).

Reply:
On a slightly related topic, the type safety here will be increased next time we update the Cleveland dependency. @Martoon-00 has recently [made some changes](https://gitlab.com/morley-framework/morley/-/merge_requests/845), and now all functions in Cleveland use a `ContractHandler cp st` instead of a `TAddress cp`.

Reply:
So I've updated `sendXtzWithAmount` with TAddress, but not `msSelfAddress`, since it is used in multiple places as `Address`, and wouldn't be useful to convert to `TAddress`.

Reply:
Sounds good :D Marking this as resolved.